File: 4_clean/inet/clean.preds.R
library(data.table)
library(tidyr)
library(scales)

clean.preds <- function(model.name){
	sizes <- c(16,32,64,128,256)
	model.files <- paste0(model.name,".",sizes,".csv")
	model.data <- lapply(model.files,read.csv,header=FALSE)
	model.data <- lapply(model.data,data.table)
	model.data <- rbindlist(model.data)
	model.data[,variable:="prob"]
	model.data[,id:=.I]
	model.data[id %% 2 == 0,variable:="pred"]
	model.data[,id:=floor((id+1)/2)]
	model.data <- spread(model.data,variable,V1)
	model.data[,sample:=rep(c(rep("valid",50000),rep("test",10000)),5)]
	model.data[,obs:=rep(c(1:50000,1:10000),5)]
	model.data[,size:=rep(sizes,each=60000)]
	model.data[,id:=NULL]
	model.data[,model:=model.name]
	return(model.data)
}

model.names <- c("resnext101_32x8d","wide_resnet101_2","efficientnet_b0","efficientnet_b7",
	"resnet101","densenet201","vgg19_bn","mobilenet_v3_large",
	"mobilenet_v3_small","googlenet","inception_v3","alexnet")

model.names <- factor(model.names,levels=model.names)

cleaned.data <- lapply(model.names,clean.preds)
cleaned.data <- rbindlist(cleaned.data)

valid.actuals <- data.table(read.csv('valid_actuals.csv',header=FALSE))
test.actuals <- data.table(read.csv('test_actuals.csv',header=FALSE))
valid.actuals[,sample:="valid"]
valid.actuals[,obs:=1:50000]
test.actuals[,sample:="test"]
test.actuals[,obs:=1:10000]

actuals <- rbind(valid.actuals,test.actuals)
setnames(actuals,c("actual","sample","obs"))

setkey(actuals,sample,obs)
setkey(cleaned.data,sample,obs)

cleaned.data <- actuals[cleaned.data]
cleaned.data[,correct:=as.numeric(actual==pred)]
saveRDS(cleaned.data,"cleaned.data.RDS")

rates <- cleaned.data[,mean(correct),by=list(sample,size,model)]
rates <- spread(rates,size,V1)
setcolorder(rates,c("sample","model","256","128","64","32","16"))

write.table(rates,"rates.txt",sep='\t',row.names=FALSE)

File: src/4_clean/inet/clean_preds.py
import pandas as pd
import numpy as np

def clean_preds(model_name):
    sizes = [16, 32, 64, 128, 256]
    model_files = [f"{model_name}.{size}.csv" for size in sizes]
    
    # Read CSV files and concatenate into a single DataFrame
    model_data = pd.concat([pd.read_csv(file, header=None) for file in model_files], ignore_index=True)
    
    # Assigning column names and creating 'variable' and 'id' columns
    model_data['variable'] = 'prob'
    model_data['id'] = model_data.index + 1
    
    # Modify 'variable' based on even/odd 'id'
    model_data.loc[model_data['id'] % 2 == 0, 'variable'] = 'pred'
    model_data['id'] = np.floor((model_data['id'] + 1) / 2).astype(int)
    
    # Spread/reshape to wide format
    model_data = model_data.pivot(index='id', columns='variable', values=0).reset_index()
    
    # Add 'sample', 'obs', 'size', and 'model' columns
    model_data['sample'] = np.repeat(np.concatenate([['valid'] * 50000, ['test'] * 10000]), 5)
    model_data['obs'] = np.tile(np.concatenate([np.arange(1, 50001), np.arange(1, 10001)]), 5)
    model_data['size'] = np.repeat(sizes, 60000)
    model_data['model'] = model_name
    
    return model_data

# Model names as in the original R script
model_names = ["resnext101_32x8d", "wide_resnet101_2", "efficientnet_b0", "efficientnet_b7",
               "resnet101", "densenet201", "vgg19_bn", "mobilenet_v3_large",
               "mobilenet_v3_small", "googlenet", "inception_v3", "alexnet"]

# Clean the data for each model and concatenate them
cleaned_data = pd.concat([clean_preds(model_name) for model_name in model_names], ignore_index=True)

# Read the actuals data
valid_actuals = pd.read_csv('valid_actuals.csv', header=None, names=['actual'])
test_actuals = pd.read_csv('test_actuals.csv', header=None, names=['actual'])

# Add 'sample' and 'obs' columns to actuals
valid_actuals['sample'] = 'valid'
valid_actuals['obs'] = np.arange(1, 50001)
test_actuals['sample'] = 'test'
test_actuals['obs'] = np.arange(1, 10001)

# Concatenate the actuals data
actuals = pd.concat([valid_actuals, test_actuals], ignore_index=True)

# Merge actuals with cleaned_data
cleaned_data = pd.merge(cleaned_data, actuals, on=['sample', 'obs'])

# Add 'correct' column based on whether the prediction matches the actual value
cleaned_data['correct'] = (cleaned_data['actual'] == cleaned_data['pred']).astype(int)

# Save the cleaned data to an RDS equivalent (Pickle in Python)
cleaned_data.to_pickle('cleaned_data.pkl')

# Calculate the mean correctness rates grouped by 'sample', 'size', and 'model'
rates = cleaned_data.groupby(['sample', 'size', 'model']).agg({'correct': 'mean'}).reset_index()

# Pivot the rates DataFrame to wide format, similar to spread in R
rates_pivot = rates.pivot(index=['sample', 'model'], columns='size', values='correct').reset_index()

# Reorder columns as per the original R script
rates_pivot = rates_pivot[['sample', 'model', 256, 128, 64, 32, 16]]

# Save the rates DataFrame to a text file
rates_pivot.to_csv('rates.txt', sep='\t', index=False)